src/token/cose/sign/sign.rs

```diff
@@ -0,0 +1,477 @@
+        'a,
+        'b,
+        B: CoseSignCipher,
+        CKP: CoseKeyProvider<'a>,
+        CAP: CoseAadProvider<'b>,
+    >(
+        self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        sig: CoseSignature,
+        payload: &[u8],
+        mut aad: &mut CAP,
+    ) -> Result<Self, CoseCipherError<B::Error>> {
+        self.try_add_detached_signature(
+            sig.clone(),
+            payload,
+            aad.lookup_aad(Some(&sig.protected.header), Some(&sig.unprotected)),
+            |tosign| {
+                sign::try_sign(
+                    backend,
+                    key_provider,
+                    Some(&sig.protected.header),
+                    Some(&sig.unprotected),
+                    tosign,
+                )
+            },
+        )
+    }
+}
+
+/// TODO for now, we assume a single successful validation implies that the validation in general is
+///      successful. However, some environments may have other policies, see
+///      https://datatracker.ietf.org/doc/html/rfc9052#section-4.1.
+pub trait CoseSignExt {
+    fn try_verify<'a, 'b, B: CoseSignCipher, CKP: CoseKeyProvider<'a>, CAP: CoseAadProvider<'b>>(
+        &self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        aad: &mut CAP,
+    ) -> Result<(), CoseCipherError<B::Error>>;
+
+    fn try_verify_detached<
+        'a,
+        'b,
+        B: CoseSignCipher,
+        CKP: CoseKeyProvider<'a>,
+        CAP: CoseAadProvider<'b>,
+    >(
+        &self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        payload: &[u8],
+        aad: &mut CAP,
+    ) -> Result<(), CoseCipherError<B::Error>>;
+}
+
+impl CoseSignExt for CoseSign {
+    fn try_verify<'a, 'b, B: CoseSignCipher, CKP: CoseKeyProvider<'a>, CAP: CoseAadProvider<'b>>(
+        &self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        mut aad: &mut CAP,
+    ) -> Result<(), CoseCipherError<B::Error>> {
+        for sigindex in 0..self.signatures.len() {
+            match self.verify_signature(
+                sigindex,
+                aad.borrow_mut().lookup_aad(
+                    Some(&self.signatures[sigindex].protected.header),
+                    Some(&self.signatures[sigindex].unprotected),
+                ),
+                |signature, toverify| {
+                    sign::try_verify(
+                        backend,
+                        key_provider,
+                        &self.signatures[sigindex].protected.header,
+                        &self.signatures[sigindex].unprotected,
+                        try_all_keys,
+                        signature,
+                        toverify,
+                    )
+                },
+            ) {
+                Ok(()) => return Ok(()),
+                Err(_) => {
+                    // TODO debug logging? Allow debugging why each verification failed?
```

Comment:
I don't get this TODO. Shouldn't we just return an error in this case?

Reply:
No, because a key with the correct key ID might not necessarily be the right key, and there could be multiple key candidiates that need to be tried (see RFC 9052, Section 7.1, especially the section regarding the keyid field), who could each have their own different error while verifying.

Probably the best solution here would be storing the errors for each verification in a `Vec` and then adding an extra error type that encapsulates all errors.

Reply:
Oh, I see, that's kind of annoying to deal with. I think what you proposed is indeed a good solution (if this is another thing that'd take more effort than I'd naively think, we can move this to a follow-up issue.)